Improve performance, add typing detection

LemonBreezes opened the pull request:
From the README:
- 0.8.2 (2025-03-15)
  Use float-times instead of timers for typing detection
- 0.8.1 (2025-03-11) 
  Add typing detection to prevent accidental chord triggering
  Add minimum delay for double-tap chords
  Performance improvements

LemonBreezes committed:
feat: Add customizable key-chord minimum delay for double-tap recognition

LemonBreezes committed:
feat: Add typing detection to prevent accidental chords

LemonBreezes committed:
fix: Ignore key-chord-one-key-min-delay during macros

LemonBreezes committed:
fix: Disable key-chord-typing-detection when recording macros

LemonBreezes committed:
fix: Improve key chord detection during macro execution

LemonBreezes committed:
style: Rename key-chord-typing-mode to in-typing-flow

LemonBreezes committed:
fix: Respect key-chord-in-macros in kbd macro execution

LemonBreezes committed:
fix: Use run-at-time for reliable typing flow detection

LemonBreezes committed:
fix: Disable key-chord-in-macros by default

LemonBreezes committed:
docs: Add customization options and typing detection docs

LemonBreezes committed:
feat: Add key tracking for performance optimization

LemonBreezes committed:
chore: Add .gitignore file

LemonBreezes committed:
style: Fix indentation and optimize typing flow check

LemonBreezes committed:
fix: Ignore non-byte characters in key-chord.el

LemonBreezes committed:
docs: Add package URL to key-chord.el

LemonBreezes committed:
docs: Update versions from 0.7.1 to 0.8.1

LemonBreezes committed:
docs: Clarify purpose of key-chord timer

LemonBreezes committed:
refactor: Replace timer-based typing detection with timestamp-based approach

LemonBreezes commented:
@BlueDrink9 @conao3 @irigone Hey, would you please check out my branch to see if it resolves your issues with key-chord? I have access to merge this PR myself but need feedback before I do so.

tarsius commented on key-chord.el:
```diff
@@ -54,6 +54,12 @@ sure when executing whether two keys were typed quickly or slowly
 when recorded.)"
   :type 'boolean)
 
+(defcustom key-chord-min-delay 0.05
```
This new variable isn't being used. I assume you start using it in a later commit. These commits should be squashed into a single commit.

LemonBreezes replied:
Hm you are somehow reviewing an old commit. Yup, I have not cleaned up the commit history. Maybe I should squash them all into 1?

tarsius commented on key-chord.el:
```diff
@@ -173,7 +179,8 @@ Commands. Please ignore that."
    ((and (not (eq first-char key-chord-last-unmatched))
          (key-chord-lookup-key (vector 'key-chord first-char)))
-    (let ((delay (if (key-chord-lookup-key
+    (let ((start-time (current-time))
```
Likewise this isn't being used (yet?).

tarsius commented:
(Lifting this to the top-level.)

> Hm you are somehow reviewing an old commit.

Those are the commits you pushed, so that's what I have to review.

> Yup, I have not cleaned up the commit history. Maybe I should squash them all into 1?

Yes, you should cleanup the commit history. I don't think squashing into a single commit is appropriate. The addition of `.gitignore` should be a separate commit for example. Metadata should be a separate commit. Cleanup such as

```diff
-  (if (/= 2 (length keys))
-      (error "Key-chord keys must have two elements"))
+  (when (/= 2 (length keys))
+    (error "Key-chord keys must have two elements"))
```

should be a separate commit. The functional change(s?) add a lot of new code, making review harder. By moving unrelated things out of the feature commit(s?) you make reviewing them easier.

Some style feedback:

- Lines should stay below 80 characters like the old code and readme.
- First line of docstrings should be a sentence ending with a period.
- Make sure not to add trailing whitespace.
- Add space between top-level forms.
- Capitalize `fix: This` consistently (or consistently not).

LemonBreezes committed:
chore: Update version to 0.8.2

LemonBreezes committed:
style: Fix line length, docstrings, and spacing

LemonBreezes committed:
docs: Fix docstring style warnings

LemonBreezes commented:
Okay. I have done the stylistic changes and squashed a few of the commits and moved some around. It's surprisingly difficult to rewrite the commit history! Even the AI can't do this well.

tarsius commented:
I cannot review this properly and because I don't use this myself I cannot even try it out to test whether it feels snappier than before. The same was true for #2 and I made the mistake of merging that, which resulted in regressions, which took a while to fix (by reverting).

I don't want to repeat that mistake. I hope we can find some users who are willing to test this for an extended period of time, before we merge.

LemonBreezes commented:
> I cannot review this properly and because I don't use this myself I cannot even try it out to test whether it feels snappier than before. The same was true for #2 and I made the mistake of merging that, which resulted in regressions, which took a while to fix (by reverting).
> 
> I don't want to repeat that mistake. I hope we can find some users who are willing to test this for an extended period of time, before we merge.

Ok. I made a Reddit post here https://www.reddit.com/r/emacs/comments/1jfa2bp/looking_for_users_to_test_a_new_version_of/. If no one takes a look there, I can post it in other places too.

LemonBreezes committed:
fix: Handle special key symbols in key-chord input method

LemonBreezes committed:
fix: Handle non-integer events in key-chord input method

LemonBreezes committed:
fix: Correct indentation in key-chord input method logic

LemonBreezes committed:
fix: Adjust indentation in key-chord.el optimization block

LemonBreezes committed:
refactor: Wrap unread events and last unmatched key in progn for consistent control flow

LemonBreezes committed:
refactor: Improve key-chord input handling with additional character validation

LemonBreezes committed:
refactor: Improve key chord second character registration and event handling

LemonBreezes committed:
refactor: Simplify key-chord special key handling and remove redundant conversions

LemonBreezes committed:
fix: Correct syntax error in key-chord.el SEARCH/REPLACE block

LemonBreezes committed:
fix: Add missing closing parentheses in key-chord-input-method function

LemonBreezes committed:
fix: Remove extra closing parenthesis in key-chord.el

LemonBreezes committed:
fix: Reset key-chord-last-unmatched after successful chord execution

LemonBreezes committed:
refactor: Reset additional key-chord state variable on mode toggle

LemonBreezes committed:
refactor: Improve key-chord typing flow and chord detection interaction

LemonBreezes committed:
style: remove trailing whitespace

LemonBreezes committed:
feat: Add read-char context handling to key-chord input method

LemonBreezes committed:
style: fix indentation in key-chord-input-method

LemonBreezes committed:
refactor: Reduce code duplication in key-chord input method

LemonBreezes commented:
Pushed some changes to fix some subtle issues with how the typing detection works as well as a subtle error. Also made it so that key-chord is completely bypassed when `read-key` is active. It just seemed like common sense because the key won't make a chord anyways, but it was causing an issue with the typing detection.

g-gundam commented:
I tried it out on Emacs 30.1 using the following configuration.
```lisp
(defun c/keybindings ()
  ;; key-chord - https://github.com/emacsorphanage/key-chord
  ;; + Chords should use both hands for maximum efficiency.
  ;;   + Practically all 2 hand chords work.
  ;;   + Some single-hand chords may work,
  ;;     but be careful not to contort your fingers.
  (key-chord-mode 1)
  (setq key-chord-two-keys-delay 0.07)
  ;; evil
  (key-chord-define evil-insert-state-map "jj" 'evil-normal-state) ; https://stackoverflow.com/a/13543550
  ;; unicode / quotes
  (key-chord-define-global "z'" "‘")      ; curly single quotes
  (key-chord-define-global "x'" "’")
  (key-chord-define-global "q'" [#x201c]) ; curly double quotes
  (key-chord-define-global "w'" [#x201d])
  ;; unicode / symbols
  (key-chord-define-global "a=" [#x0fd5]) ;
  (key-chord-define-global "b=" [#x0fd6]) ;
  (key-chord-define-global "c0" [#x00a9]) ; © copyright
  (key-chord-define-global "y=" [#x00a5]) ; ¥ yen
  (key-chord-define-global "l=" [#x00a3]) ; £ pound
  (key-chord-define-global "3." [#x2764]) ; ❤ heart
  ;; unicode / letters
  (key-chord-define-global "n`" [#x00f1]) ; ñ enye
  (key-chord-define-global "a'" [#xe1])   ; á
  (key-chord-define-global "e'" [#xe9])   ; é
  (key-chord-define-global "i'" [#xed])   ; í
  (key-chord-define-global "o'" [#xf3])   ; ó
  ;; julia programming
  (key-chord-define-global "z\\" [#x7c #x3e]) ; |>
  ;; org-mode
  (key-chord-define-global "t0" 'i/org-time-stamp-inactive)
  (key-chord-define-global "28" (u/org-insert-datestamped-heading-fn 2))
  (key-chord-define-global "38" (u/org-insert-datestamped-heading-fn 3))
  )

(use-package key-chord
  :vc (:url "https://github.com/LemonBreezes/key-chord" :branch "master")
  :config
  (c/keybindings)
  ;; key-chord international hack
  (advice-add
   'toggle-input-method
   :after
   (lambda (&rest r)
     (if (equal current-input-method nil)
	 (progn
	   (c/keybindings)
	   (message "rebinding"))
       (message current-input-method)))
   '((name . keychords))))
```
The only difference from my normal config is the use-package where instead of `:vc` it normally says `:ensure t`. 

##  I did not notice any regressions.

- Everything in my config worked as expected.
- I didn't really notice any differences in feel.

PS:  This way of setting up key-chord is my way of working around the [limitations](https://github.com/emacsorphanage/key-chord/?tab=readme-ov-file#limitations) mentioned in the README about key-chord and internationalization packages not working well together.  The advice that's added can help them coexist.  While using international input, you can't use key-chord, but when you toggle back to regular input, you get all your key-chord config back again.

auslegungssache commented:
Hi! Really liking the changes you've made :>
It makes chording feel more reliable and native!!!

However, I DID notice one regression.
It seems to break `use-package-chords` (at least under doom emacs). The keys don't get bound until one goes and binds another key with the same prefix normally.
I honestly can't explain to myself why this is the case, but it doesn't happen with the normal `key-chord` package

EDIT:
Forgot to include a screenshot of my config

![image](https://github.com/user-attachments/assets/ccb1fd1c-db1b-435e-bcab-a83bdfc2a7de)

LemonBreezes committed:
refactor: Improve key-chord timing check with elapsed time variable

LemonBreezes commented:
> Hi! Really liking the changes you've made :> It makes chording feel more reliable and native!!!
> 
> However, I DID notice one regression. It seems to break `use-package-chords` (at least under doom emacs). The keys don't get bound until one goes and binds another key with the same prefix normally. I honestly can't explain to myself why this is the case, but it doesn't happen with the normal `key-chord` package
> 
> EDIT: Forgot to include a screenshot of my config
> 
> ![image](https://private-user-images.githubusercontent.com/23395482/427613433-ccb1fd1c-db1b-435e-bcab-a83bdfc2a7de.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDMxMTY0MDAsIm5iZiI6MTc0MzExNjEwMCwicGF0aCI6Ii8yMzM5NTQ4Mi80Mjc2MTM0MzMtY2NiMWZkMWMtZGIxYi00MzVlLWJjYWItYTgzYmRmYzJhN2RlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzI3VDIyNTUwMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTIxOGM4Y2FhNTBmNmQ0ZjY3ZjI5NWExOGRkMjU5OGEzNDMxZTFkYjJhYTc4YmJlMjFmODA0MWFkNGRlZTVmODMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.2hnCk4xEQbCsgTCNvGAPh3jebZgEC6A2p1VmY5ovUXY)

Okay. I have identified what the cause of this issue is and am working on a fix. This will most likely involve a PR to `use-package-chords` (or `use-package` if it's just a part of `use-package` now).

LemonBreezes committed:
feat: Add functions to register/unregister chord keys

LemonBreezes committed:
fix: Always track keys used in chords

LemonBreezes committed:
refactor: Make key registration/unregistration unconditional

LemonBreezes commented:
> > Hi! Really liking the changes you've made :> It makes chording feel more reliable and native!!!
> > However, I DID notice one regression. It seems to break `use-package-chords` (at least under doom emacs). The keys don't get bound until one goes and binds another key with the same prefix normally. I honestly can't explain to myself why this is the case, but it doesn't happen with the normal `key-chord` package
> > EDIT: Forgot to include a screenshot of my config
> > ![image](https://private-user-images.githubusercontent.com/23395482/427613433-ccb1fd1c-db1b-435e-bcab-a83bdfc2a7de.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDMxMTY0MDAsIm5iZiI6MTc0MzExNjEwMCwicGF0aCI6Ii8yMzM5NTQ4Mi80Mjc2MTM0MzMtY2NiMWZkMWMtZGIxYi00MzVlLWJjYWItYTgzYmRmYzJhN2RlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzI3VDIyNTUwMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTIxOGM4Y2FhNTBmNmQ0ZjY3ZjI5NWExOGRkMjU5OGEzNDMxZTFkYjJhYTc4YmJlMjFmODA0MWFkNGRlZTVmODMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.2hnCk4xEQbCsgTCNvGAPh3jebZgEC6A2p1VmY5ovUXY)
> 
> Okay. I have identified what the cause of this issue is and am working on a fix. This will most likely involve a PR to `use-package-chords` (or `use-package` if it's just a part of `use-package` now).

Okay. I have made a PR in https://github.com/jwiegley/use-package/pull/1088 to fix this. In the meantime, you can add 
```elisp
(key-chord-register-keys ?' ?a)
(key-chord-register-keys ?' ?s)
(key-chord-register-keys ?' ?3)
(key-chord-register-keys ?' ?2)
(key-chord-register-keys ?' ?1)
```
to your configuration.

tarsius commented:
> I have made a PR in https://github.com/jwiegley/use-package/pull/1088 to fix this.

`use-package` isn't being maintained in that repository anymore. It is maintained in Emacs' repository.

auslegungssache commented:
I might be wrong here but I think that's the wrong package...
Since use-package got merged into emacs, they dropped the key chord support. To get it back, you need to use this external package [use-package-chords](https://github.com/waymondo/use-package-chords)

LemonBreezes commented:
> I might be wrong here but I think that's the wrong package...
> Since use-package got merged into emacs, they dropped the key chord support. To get it back, you need to use this external package [use-package-chords](https://github.com/waymondo/use-package-chords)

Ohh, I had submitted an Emacs bug report too. I'll also open a PR there. 
EDIT: https://github.com/waymondo/use-package-chords/pull/14

LemonBreezes commented:
Ok. The `use-package-chords` PR has been merged so now that is compatible.

tarsius commented:
Please feel free to merge once you think this is ready to go.

LemonBreezes committed:
refactor: Remove unused key-chord-update-keys-in-use function

Konubinix commented:
Tried for one week and still did not notice any issue. I actually forgot about it until now.
